File: core/src/main/java/com/lomekwi/cave/resource/media/FontRes.java
package com.lomekwi.cave.resource.media;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator;
import com.badlogic.gdx.utils.IntMap;
import com.lomekwi.cave.resource.Resource;

import java.io.Serial;
import java.io.Serializable;

public class FontRes implements Resource, Serializable {
    private final String path;
    private transient FreeTypeFontGenerator generator;
    private transient IntMap<BitmapFont> fontCache;

    @Serial
    private static final long serialVersionUID = 1L;

    public FontRes(String path) {
        this.path = path;
    }

    public BitmapFont getFont(int size) {
        if (fontCache == null) fontCache = new IntMap<>();
        BitmapFont cached = fontCache.get(size);
        if (cached != null) return cached;
        if (generator == null) {
            var handle = Gdx.files.internal(path);
            if (!handle.exists()) {
                handle = Gdx.files.absolute(path);
            }
            generator = new FreeTypeFontGenerator(handle);
        }
        FreeTypeFontGenerator.FreeTypeFontParameter param =
            new FreeTypeFontGenerator.FreeTypeFontParameter();
        param.size = size;
        param.incremental = true;
        BitmapFont font = generator.generateFont(param);
        fontCache.put(size, font);
        return font;
    }

    public String getPath() {
        return path;
    }

    @Override
    public void close() {
        if (fontCache != null) {
            for (BitmapFont font : fontCache.values()) {
                font.dispose();
            }
            fontCache.clear();
        }
        if (generator != null) {
            generator.dispose();
            generator = null;
        }
    }
}

File: core/src/main/scala/com/lomekwi/cave/resource/media/FontRes.scala
package com.lomekwi.cave.resource.media

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator
import com.badlogic.gdx.utils.IntMap
import com.lomekwi.cave.resource.Resource

import java.io.Serializable
import scala.compiletime.uninitialized


@SerialVersionUID(1L)
class FontRes(private val path: String) extends Resource with Serializable {
  @transient private var generator: FreeTypeFontGenerator = uninitialized
  @transient private lazy val fontCache: IntMap[BitmapFont] = new IntMap[BitmapFont]()

  def getFont(size: Int): BitmapFont = {
    Option(fontCache.get(size)).getOrElse {
      val param = new FreeTypeFontGenerator.FreeTypeFontParameter()
      param.size = size
      param.incremental = true
      val font = requireGenerator().generateFont(param)
      fontCache.put(size, font)
      font
    }
  }

  private def requireGenerator(): FreeTypeFontGenerator = {
    if (generator == null) {
      var handle = Gdx.files.internal(path)
      if (!handle.exists()) {
        handle = Gdx.files.absolute(path)
      }
      generator = new FreeTypeFontGenerator(handle)
    }
    generator
  }

  def getPath: String = {
    path
  }

  override def close(): Unit = {
    val fontIt = fontCache.values().iterator()
    while (fontIt.hasNext) {
      fontIt.next().dispose()
    }
    fontCache.clear()
    if (generator != null) {
      generator.dispose()
      generator = null
    }
  }
}
